### PR TITLE
Refactor initialization for simpler maintenance

### PR DESCRIPTION
**Pull request by coroa**
# Please confirm that this PR has done the following:

- [ ] Tests Added
- [ ] ~Documentation Added~
- [ ] ~Name of contributors Added to AUTHORS.rst~
- [x] Description in RELEASE_NOTES.md Added

# Description of PR

~On-top of #729.~ Rebased to `main`.

Splits `utils.format_data` into 6 different functions:
1. `_convert_r_columns(df)` - Check and convert R-style year columns
2. `_knead_data(df, **kwargs)` - Replace, rename and concat according to user arguments
3. `_format_from_database(df)` - Post-process database results
4. `_intuit_column_groups(df, index)` - Check and categorise columns in dataframe
5. `_format_data_to_series(df, index)` - Convert a long or wide pandas dataframe to a series with the required columns
6. `_validate_complete_index(df)`

Functionally it should be neutral. Make sure to look at individual commits to follow the refactoring trail.

### CODE REVIEW

**Comment by gidden** [comment]
This LGTM but best if @danielhuppmann approves!

File: pyam/utils.py [hunk 1]
```diff
@@ -181,13 +176,9 @@ def read_file(path, *args, **kwargs):
...
-def format_data(df, index, **kwargs):
-    """Convert a pandas.Dataframe or pandas.Series to the required format"""
-    if isinstance(df, pd.Series):
-        df.name = df.name or "value"
-        df = df.to_frame()
+def _convert_r_columns(df):
+    """Check and convert R-style year columns"""
 
-    # check for R-style year columns, converting where necessary
     def convert_r_columns(c):
```
Comment (danielhuppmann) [inline pyam/utils.py:182]: Repeating the function name within is a bit confusing, propose to be more explicit.
```suggestion
    def strip_R_integer_prefix(c):
```

File: pyam/utils.py [hunk 2]
```diff
@@ -204,7 +195,11 @@ def convert_r_columns(c):
             pass
         return c
 
-    df.columns = df.columns.map(convert_r_columns)
+    return df.set_axis(df.columns.map(convert_r_columns), axis="columns")
```
Comment (danielhuppmann) [inline pyam/utils.py:198]:
```suggestion
    return df.set_axis(df.columns.map(strip_R_integer_prefix), axis="columns")
```

File: pyam/utils.py [hunk 3]
```diff
@@ -236,26 +231,26 @@ def convert_r_columns(c):
...
-            df.loc[:, "model"] = scen.apply(lambda s: s.split("-")[0].strip())
-            df.loc[:, "scenario"] = scen.apply(
-                lambda s: "-".join(s.split("-")[1:]).strip()
-            )
 
-    # reset the index if meaningful entries are included there
-    if not list(df.index.names) == [None]:
-        df.reset_index(inplace=True)
+def _format_from_database(df):
+    """Post-process database results"""
```
Comment (danielhuppmann) [inline pyam/utils.py:238]:
```suggestion
def _format_from_legacy_database(df):(df):
    """Process data from legacy databases (SSP and earlier)"""
```

File: pyam/utils.py [hunk 4]
```diff
@@ -323,23 +306,67 @@ def convert_r_columns(c):
...
+    df = _convert_r_columns(df)
+
+    if kwargs:
+        df = _knead_data(df, **kwargs)
+
+    # all lower case
+    df.rename(columns={c: str(c).lower() for c in df.columns if isstr(c)}, inplace=True)
+
+    if "notes" in df.columns:  # this came from the database
+        df = _format_from_database(df)
```
Comment (danielhuppmann) [inline pyam/utils.py:364]:
```suggestion
    if "notes" in df.columns:  # this came from a legacy database (SSP or earlier)
        df = _format_from_legacy_database(df)
```

File: pyam/utils.py [hunk 5]
```diff
@@ -323,23 +306,67 @@ def convert_r_columns(c):
...
+    elif not list(df.index.names) == [None]:
+        # reset the index if meaningful entries are included there
+        df = df.reset_index()
+
+    df = _convert_r_columns(df)
+
+    if kwargs:
+        df = _knead_data(df, **kwargs)
+
+    # all lower case
```
Comment (danielhuppmann) [inline pyam/utils.py:360]:
```suggestion
    # cast all columns names to lower case
```

File: pyam/utils.py [hunk 6]
```diff
@@ -323,23 +306,67 @@ def convert_r_columns(c):
...
+    df = df.assign(unit=df["unit"].fillna(""))
 
-        # cast to long format, set
-        df.set_index(index + REQUIRED_COLS + extra_cols, inplace=True)
-        df = df.stack(dropna=True)
-        df.name = "value"
-        df.index.names = df.index.names[:-1] + [time_col]
+    df, time_col, extra_cols = _format_data_to_series(df, index)
 
     # cast value column to numeric
```
Comment (danielhuppmann) [inline pyam/utils.py:371]: Might also put these two checks into their own function `_check_data_integrity()`...

File: pyam/utils.py [hunk 7]
```diff
@@ -236,26 +231,26 @@ def convert_r_columns(c):
...
-                lambda s: "-".join(s.split("-")[1:]).strip()
-            )
 
-    # reset the index if meaningful entries are included there
-    if not list(df.index.names) == [None]:
-        df.reset_index(inplace=True)
+def _format_from_database(df):
+    """Post-process database results"""
+
+    logger.info("Ignoring notes column in dataframe")
```
Comment (danielhuppmann) [inline pyam/utils.py:240]: For consistency with other messages, e.g., the error "Duplicate rows in `data`"
```suggestion
    logger.info("Ignoring notes column in `data`")
```

File: pyam/utils.py [hunk 8]
```diff
@@ -236,26 +231,26 @@ def convert_r_columns(c):
...
-            # model and scenario are jammed together in RCP data
-            scen = df["scenario"]
-            df.loc[:, "model"] = scen.apply(lambda s: s.split("-")[0].strip())
-            df.loc[:, "scenario"] = scen.apply(
-                lambda s: "-".join(s.split("-")[1:]).strip()
-            )
+    return df
+
+
+def _format_from_legacy_database(df):(df):
```
Comment (gidden) [inline pyam/utils.py:237]:
```suggestion
def _format_from_legacy_database(df):
```

**Comment by gidden** [comment]
LGTM - thanks @coroa ! Will merge after tests pass.

**Pull request merged**